### PR TITLE
Varadski add timestamp to message

### LINKED ISSUE #20: Add timestamp to Messages
A `Message` should have a timestamp of when it was created, but currently it does not. `SendMessageCommand` and `MessageSentEvent` need to be updated appropriately as well.

### PR DESCRIPTION
**Pull request by Varadski**
Resolved #20.

### CODE REVIEW

File: domain/src/test/java/eu/execom/buildingmanager/domain/conversation/ConversationCommandHandlerTest.java [hunk 1]
```diff
@@ -34,19 +33,21 @@ public void setUp() {
...
+    Message message = MessageMother.message("Hello.", tenant, buildingManager, timestamp);
 
     StartConversationCommand startConversationCommand = new StartConversationCommand(buildingManager, tenant, message);
-    ConversationStartedEvent conversationStartedEvent = new ConversationStartedEvent(id, buildingManager, tenant,
-        message);
+    ConversationStartedEvent conversationStartedEvent = new ConversationStartedEvent(id, buildingManager, tenant);
+    MessageSentEvent messageSentEvent = new MessageSentEvent(id, message.getText(), tenant, buildingManager, timestamp);
 
-    conversationFixture.givenNoPriorActivity().when(startConversationCommand).expectEvents(conversationStartedEvent);
+    conversationFixture.givenNoPriorActivity()
```
[the remark this answers is not in the document]
Comment (Varadski) [reply domain/src/test/java/eu/execom/buildingmanager/domain/conversation/ConversationCommandHandlerTest.java]: This behaviour is defined by the formatter.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java [hunk 2]
```diff
@@ -34,43 +36,46 @@
...
     Tenant tenant = command.getTenant();
-    Message message = command.getMessage();
+    String messageText = command.getMessageText();
+    User messageAuthor = command.getMessageAuthor();
+    LocalDateTime createdAt = creationTimeGenerator.generateTimestamp();
 
     Assert.notNull(buildingManager, "Missing building manager");
     Assert.notNull(tenant, "Missing tenant");
-    Assert.isTrue(message.isValid(), "The message isn't valid");
+    Assert.hasLength(messageText, "<Missing message text");
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: Please remove < from the message.

File: domain/src/main/java/eu/execom/buildingmanager/domain/utility/CreationTimeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package eu.execom.buildingmanager.domain.utility;
+
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+
+@Component
+public class CreationTimeGenerator {
+  public LocalDateTime generateTimestamp() {
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/utility/CreationTimeGenerator.java]: Please rename to `now()`.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java [hunk 4]
```diff
@@ -34,43 +36,46 @@
...
-  public Conversation(StartConversationCommand command, IdGenerator idGenerator) {
+  public Conversation(StartConversationCommand command, IdGenerator idGenerator,
+      TimeProvider timeProvider) {
     String id = idGenerator.generateId();
     BuildingManager buildingManager = command.getBuildingManager();
     Tenant tenant = command.getTenant();
-    Message message = command.getMessage();
+    String messageText = command.getMessageText();
+    User messageAuthor = command.getMessageAuthor();
+    LocalDateTime createdAt = timeProvider.now();
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: You should create this variable where it is used, and that is in the handler of the `MessageSentEvent`.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java [hunk 5]
```diff
@@ -34,43 +36,46 @@
...
   }
 
   @CommandHandler
-  public void on(SendMessageCommand command) {
+  public void on(SendMessageCommand command, TimeProvider timeProvider) {
     String text = command.getText();
-    User sender = command.getSender();
-    User receiver = command.getReceiver();
+    User author = command.getAuthor();
+    LocalDateTime timestamp = timeProvider.now();
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: Move this to the handler of the `MessageSentEvent`.

File: domain/src/main/java/eu/execom/buildingmanager/domain/utility/TimeProvider.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package eu.execom.buildingmanager.domain.utility;
+
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+
+@Component
+public class TimeProvider {
+  public LocalDateTime now() {
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/utility/TimeProvider.java]: Add one line above to make the class more readable.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java (hunk 4 above)
Comment (Varadski) [reply domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: `timeProvider.now()` generates the current UTC time. During aggregate reconstitution the time provided will be inaccurate in it is generated within the event handler, as opposed to being stored in the event itself.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java (hunk 5 above)
Comment (Varadski) [reply domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: `timeProvider.now()` generates the current UTC time. During aggregate reconstitution the time provided will be inaccurate in it is generated within the event handler, as opposed to being stored in the event itself.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java (hunk 4 above)
Comment (stanislavgrujic) [reply domain/src/main/java/eu/execom/buildingmanager/domain/conversation/Conversation.java]: Good catch! :)

**Comment by Varadski** [comment]
Resolved #28.

File: domain/src/main/java/eu/execom/buildingmanager/domain/conversation/StartConversationSaga.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import static org.axonframework.eventhandling.saga.SagaLifecycle.end;
+
+@Saga
+public class StartConversationSaga {
+
+  @Inject
+  private transient CommandGateway commandGateway;
+
+  private int commandCounter = 0;
```
Comment (stanislavgrujic) [inline domain/src/main/java/eu/execom/buildingmanager/domain/conversation/StartConversationSaga.java]: You're actually counting received events, not commands. Please use proper naming. :)

**Pull request merged**